op-proposer/proposer/l2_output_submitter.go

```diff
@@ -212,75 +220,244 @@ func NewL2OutputSubmitterWithSigner(
+	}
+
+	// Always propose if it's part of the Finalized L2 chain. Or if allowed, if it's part of the safe L2 chain.
+	if !(output.BlockRef.Number <= output.Status.FinalizedL2.Number || (l.allowNonFinalized && output.BlockRef.Number <= output.Status.SafeL2.Number)) {
+		l.log.Debug("not proposing yet, L2 block is not ready for proposal",
+			"l2_proposal", output.BlockRef,
+			"l2_safe", output.Status.SafeL2,
+			"l2_finalized", output.Status.FinalizedL2,
+			"allow_non_finalized", l.allowNonFinalized)
+		return nil, false, nil
+	}
+	return output, true, nil
 }
 
-// dialRollupClientWithTimeout attempts to dial the RPC provider using the provided
-// URL. If the dial doesn't complete within defaultDialTimeout seconds, this
-// method will return an error.
-func dialRollupClientWithTimeout(ctx context.Context, url string) (*sources.RollupClient, error) {
-	ctxt, cancel := context.WithTimeout(ctx, defaultDialTimeout)
-	defer cancel()
+// CreateProposalTx transforms an output response into a signed output transaction.
+// It does not send the transaction to the transaction pool.
+func (l *L2OutputSubmitter) CreateProposalTx(ctx context.Context, output *eth.OutputResponse) (*types.Transaction, error) {
+	nonce, err := l.l1Client.NonceAt(ctx, l.from, nil)
+	if err != nil {
+		l.log.Error("Failed to get nonce", "err", err, "from", l.from)
+		return nil, err
+	}
 
-	rpcCl, err := rpc.DialContext(ctxt, url)
+	opts := &bind.TransactOpts{
+		From: l.from,
+		Signer: func(addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+			return l.signerFn(ctx, addr, tx)
+		},
+		Context: ctx,
+		Nonce:   new(big.Int).SetUint64(nonce),
+		NoSend:  true,
+	}
+
+	tx, err := l.l2ooContract.ProposeL2Output(
+		opts,
+		output.OutputRoot,
+		new(big.Int).SetUint64(output.BlockRef.Number),
+		output.Status.CurrentL1.Hash,
+		new(big.Int).SetUint64(output.Status.CurrentL1.Number))
 	if err != nil {
+		l.log.Error("failed to create the ProposeL2Output transaction", "err", err)
 		return nil, err
 	}
+	return tx, nil
+}
 
-	return sources.NewRollupClient(client.NewBaseRPCClient(rpcCl)), nil
+// SendTransactionExt sends a transaction through the transaction manager which handles automatic
+// price bumping.
+// It also hardcodes a timeout of 100s.
+func (l *L2OutputSubmitter) SendTransactionExt(ctx context.Context, tx *types.Transaction) error {
+	// Construct the a closure that will update the txn with the current gas prices.
+	nonce := tx.Nonce()
+	updateGasPrice := func(ctx context.Context) (*types.Transaction, error) {
+		l.log.Info("proposer updating batch tx gas price", "nonce", nonce)
+		return l.UpdateGasPrice(ctx, tx)
+	}
+
+	// Wait until one of our submitted transactions confirms. If no
+	// receipt is received it's likely our gas price was too low.
+	cCtx, cancel := context.WithTimeout(ctx, 100*time.Second)
+	defer cancel()
+	receipt, err := l.txMgr.Send(cCtx, updateGasPrice, l.SendTransaction)
```

Comment:
Properly set a timeout b/c if gas estimation fails inside the txmanager, it hangs forever.